bootstrap/pkg/client/gcp/gcp.go

```diff
@@ -802,13 +790,87 @@ func (gcp *Gcp) createSecrets(options map[string]interface{}) error {
 			string(kftypes.OAUTH_SECRET), CLIENT_SECRET)
 	}
-
-	return gcp.writeGcpSecret(k8sClient, KUBEFLOW_OAUTH, map[string][]byte{
+	return gcp.writeGcpSecret(client, KUBEFLOW_OAUTH, map[string][]byte{
 		strings.ToLower(CLIENT_ID):     []byte(oauthId),
 		strings.ToLower(CLIENT_SECRET): []byte(oauthSecret),
 	})
 }
 
+func (gcp *Gcp) createBasicAuthSecret(client *clientset.Clientset, options map[string]interface{}) error {
+	if !gcp.GcpApp.Spec.UseBasicAuth {
+		log.Infof("Not using basic auth, skip creating basic auth login secret.")
+		return nil
+	}
+	username := ""
+	if options[string(kftypes.BASIC_AUTH_USERNAME)] != nil &&
+		options[string(kftypes.BASIC_AUTH_USERNAME)].(string) != "" {
+		username = options[string(kftypes.BASIC_AUTH_USERNAME)].(string)
+	} else {
+		username = os.Getenv(BASIC_AUTH_USERNAME)
+	}
+	if username == "" {
+		return fmt.Errorf("At least one of --%v or ENV `%v` needs to be set.",
+			string(kftypes.BASIC_AUTH_USERNAME), BASIC_AUTH_USERNAME)
+	}
+
+	password := ""
+	if options[string(kftypes.BASIC_AUTH_PASSWORD)] != nil &&
+		options[string(kftypes.BASIC_AUTH_PASSWORD)].(string) != "" {
+		password = options[string(kftypes.BASIC_AUTH_PASSWORD)].(string)
+	} else {
+		password = os.Getenv(BASIC_AUTH_PASSWORD)
+	}
+	if password == "" {
+		return fmt.Errorf("At least one of --%v or ENV `%v` needs to be set.",
+			string(kftypes.BASIC_AUTH_PASSWORD), BASIC_AUTH_PASSWORD)
+	}
+	passwordHash, err := bcrypt.GenerateFromPassword([]byte(password), 10)
+	if err != nil {
+		return fmt.Errorf("error when hashing password: %v", err)
+	}
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      BASIC_AUTH_SECRET,
+			Namespace: gcp.GcpApp.Namespace,
+		},
+		Data: map[string][]byte{
+			"username":     []byte(username),
+			"passwordhash": []byte(base64.StdEncoding.EncodeToString(passwordHash)),
+		},
+	}
+	_, err = client.CoreV1().Secrets(gcp.GcpApp.Namespace).Update(secret)
+	if err != nil {
+		log.Warnf("Updating basic auth login is failed, trying to create one: %v", err)
+		_, err = client.CoreV1().Secrets(gcp.GcpApp.Namespace).Create(secret)
+	}
+	return err
+}
+
+func (gcp *Gcp) createSecrets(options map[string]interface{}) error {
+	ctx := context.Background()
+	k8sClient, err := gcp.getK8sClientset(ctx)
+	if err != nil {
+		return fmt.Errorf("Get K8s clientset error: %v", err)
+	}
+	adminEmail := getSA(gcp.GcpApp.Name, "admin", gcp.GcpApp.Spec.Project)
+	userEmail := getSA(gcp.GcpApp.Name, "user", gcp.GcpApp.Spec.Project)
+	if err := gcp.createGcpServiceAcctSecret(ctx, k8sClient, adminEmail, ADMIN_SECRET_NAME); err != nil {
+		return fmt.Errorf("cannot create admin secret %v Error %v", ADMIN_SECRET_NAME, err)
+
+	}
+	if err := gcp.createGcpServiceAcctSecret(ctx, k8sClient, userEmail, USER_SECRET_NAME); err != nil {
+		return fmt.Errorf("cannot create user secret %v Error %v", USER_SECRET_NAME, err)
+
+	}
+	if err := gcp.createIapSecret(ctx, k8sClient, options); err != nil {
```

Comment:
shouldn't this be done conditionally (!UseBasicAuth?)
same as below

Reply:
metric-controller also uses it.  added a comment as TODO such that we should check not only `UseBasicAuth`, but also if `metric-controller` is part of components.

Reply:
also, `UseBasicAuth` checks is in `createBasicAuthSecret`

Reply:
Should be safe to remove IapSecret when UseBasicAuth.
I'll change metric-controller to skip IapSecret when not available.

Reply:
Done.